src/snowflake/connector/config_manager.py

```diff
@@ -0,0 +1,333 @@
+from typing import Any, Callable, TypeVar
+from warnings import warn
+
+import tomlkit
+from tomlkit.items import Table
+from typing_extensions import Literal
+
+from snowflake.connector.constants import CONFIG_FILE
+from snowflake.connector.errors import ConfigManagerError, ConfigSourceError
+
+_T = TypeVar("_T")
+
+LOGGER = logging.getLogger(__name__)
+READABLE_BY_OTHERS = stat.S_IRGRP | stat.S_IROTH
+
+
+class ConfigOption:
+    """ConfigOption represents a flag/setting.
+
+    The class knows how to read the value out of all sources and implements
+    order of precedence between them.
+
+    Attributes:
+        name: Name of this ConfigOption.
+        parse_str: A function that can turn str to the desired type, useful
+          for reading value from environmental variable.
+        choices: An iterable of all possible values that are allowed for
+          this option.
+        env_name: Environmental variable value should be read from, if not
+          supplied, we'll construct this. False disables reading from
+          environmental variable, None uses the auto generated variable name
+          and explicitly provided string overwrites the default one.
+        _root_parser: Reference to the root parser. Used to efficiently
+          refer to cached config file. Is supplied by the parent
+          ConfigManager.
+        _nest_path: The names of the ConfigManagers that this option is
+          nested in. Used be able to efficiently resolve where to grab
+          value out of configuration file and construct environment
+          variable name. This is supplied by the parent ConfigManager.
+    """
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        _root_parser: ConfigManager,
+        _nest_path: list[str],
+        parse_str: Callable[[str], _T] | None = None,
+        choices: Iterable[Any] | None = None,
+        env_name: str | None | Literal[False] = None,
+    ) -> None:
+        """Create a config option that can read values from different locations.
+
+        Args:
+            name: Name to assign to this ConfigOption.
+            parse_str: String parser function for this ConfigOption.
+            choices: List of possible values for this ConfigOption.
+            env_name: Environmental variable name value should be read from.
+            _root_parser: Reference to the root parser. Should be supplied by
+              the parent ConfigManager.
+            _nest_path: The names of the ConfigManagers that this option is
+              nested in. This is supplied by the parent ConfigManager.
+        """
+        self.name = name
+        self.parse_str = parse_str
+        self.choices = choices
+        self._nest_path = _nest_path + [name]
+        self._root_parser: ConfigManager = _root_parser
+        self.env_name = env_name
+
+    def value(self) -> Any:
+        """Retrieve a value of option.
+
+        This function implements order of precedence between different sources.
```

Comment:
Could it be a property?

Reply:
I explained my reasoning here in another comment https://github.com/snowflakedb/snowflake-connector-python/pull/1549#discussion_r1187962914